trie/fog.py

```diff
@@ -0,0 +1,284 @@
+            raise ValidationError(
+                f"Got duplicate sub_segments in {sub_segments} to HexaryTrieFog.explore()"
+            )
+
+        # Add further validation that no segment is a prefix of another
+        all_lengths = set(len(segment) for segment in sub_segments)
+        if len(all_lengths) > 1:
+            for segment in sub_segments:
+                shorter_lengths = [length for length in all_lengths if length < len(segment)]
+                for check_length in shorter_lengths:
+                    trimmed_segment = segment[:check_length]
+                    if trimmed_segment in sub_segments:
+                        raise ValidationError(
+                            f"Cannot add {segment} which is a child of segment {trimmed_segment}"
+                        )
+
+        new_fog_prefixes.update([old_prefix + segment for segment in sub_segments])
+        return self._new_trie_fog(new_fog_prefixes)
+
+    def mark_all_complete(self, prefixes: Sequence[NibblesInput]) -> 'HexaryTrieFog':
+        """
+        These might be leaves, or prefixes with 0 unknown keys within the range.
+
+        This is equivalent to the following, but with better performance:
+
+            result_fog = old_fog
+            for complete_prefix in prefixes:
+                result_fog = result_fog.explore(complete_prefix, ())
+        """
+        new_unexplored_prefixes = self._unexplored_prefixes.copy()
+        for prefix in prefixes:
+            if prefix not in new_unexplored_prefixes:
+                raise ValidationError(
+                    f"When marking {prefix} complete, could not find in {new_unexplored_prefixes!r}"
+                )
+
+            new_unexplored_prefixes.remove(prefix)
+        return self._new_trie_fog(new_unexplored_prefixes)
+
+    def nearest_unknown(self, key_input: NibblesInput) -> Nibbles:
+        """
+        Find the foggy prefix that is nearest to the supplied key.
+
+        :raises IndexError: if there are no foggy prefixes remaining
+        """
+        if isinstance(key_input, Nibbles):
+            key = key_input
+        else:
+            key = Nibbles(key_input)
+
+        # This will raise an IndexError if no prefixes exist
+        index = self._unexplored_prefixes.bisect(key)
+
+        if index == 0:
+            return self._unexplored_prefixes[0]
+        elif index == len(self._unexplored_prefixes):
+            return self._unexplored_prefixes[-1]
+        else:
+            nearest_left = self._unexplored_prefixes[index - 1]
+            nearest_right = self._unexplored_prefixes[index]
+
+            # is the left or right unknown prefix closer?
+            left_distance = self._prefix_distance(nearest_left, key)
+            right_distance = self._prefix_distance(key, nearest_right)
+            if left_distance < right_distance:
+                return nearest_left
+            else:
+                return nearest_right
```

Comment:
Looks like this will *prefer* the key to the right over the key to the left if their respective distance is the same.  Maybe worth noting this in the docstring.